packages/flutter_tools/lib/src/debug_adapters/flutter_adapter.dart

```diff
@@ -350,7 +351,16 @@ class FlutterDebugAdapter extends FlutterBaseDebugAdapter with VmServiceInfoFile
     // Flutter requests are always wrapped in brackets as an array.
     final String payload = '[$messageString]\n';
     _logTraffic('==> [Flutter] $payload');
-    process.stdin.writeln(payload);
+    ProcessUtils.writelnToStdinGuarded(
+      stdin: process.stdin,
+      line: payload,
+      onError: (Object error, StackTrace stackTrace) {
+        throw DebugAdapterException(
+          'Writing "${payload.trim()}" to a Flutter daemon sub-process failed:\n'
+          '$error\n\n$stackTrace\n',
+        );
```

Comment:
@christopherfujino, does `DebugAdapaterException` get caught somewhere upstream? Otherwise, doesn't this not solve the root issue of the crash (though it does give us a better stack trace and error message)?

The heart of this question is that there are places in tool code where we call `writeln` on `Process::stdin`, and I imagine that silencing them with `onErrors` that produce `ToolExit`s may silence legitimate crashes. I can find examples upon request (compile.dart would be a start).

Reply:
> Otherwise, doesn't this not solve the root issue of the crash (though it does give us a better stack trace and error message)?

I don't know the root issue of the crash, so I don't know how to solve it. If it's clear to you the root issue, by all means, please fix it :)

Reply:
Discussed and resolved on Discord: https://discord.com/channels/608014603317936148/608022056616853515/1200554396166144010

I'll go over all our calls of `writeln` on `Process::stdin` and examine each on a case by case to determine whether or not we should swallow exceptions or throw our own.

Normally, I up would split up work into smaller PRs rather than trying to solve the parent Issue in one go, but this would make cherry-picking more work (assuming the risk of needing to revert is indeed low).